### PR TITLE
credstash: add standalone Python application

### LINKED ISSUE #47751: Credstash not runnable as an application
## Issue description

After installing [credstash](https://github.com/fugue/credstash) with nix-env, running it raises a SyntaxError. It looks like Python is attempting to import a module from one of the wrapper bash scripts.

The bin directory has shell scripts `credstash` and `credstash.py`, each of which `exec`s actual python code held in `.credstash-wrapped` and `.credstash.py-wrapped`. In `.credstash-wrapped`, it tries to load a Python module from the `credstash.py` shell wrapper, causing a `SyntaxError`.

```
$ nix-env --install --attr nixos.python27Packages.credstash
$ credstash setup
Traceback (most recent call last):
  File "/nix/store/rd4l1qkpx4qfzamh08a701nxr34kgjzr-python2.7-credstash-1.14.0/bin/.credstash-wrapped", line 8, in <module>
    from credstash import main
  File "/nix/store/rd4l1qkpx4qfzamh08a701nxr34kgjzr-python2.7-credstash-1.14.0/bin/credstash.py", line 2
    export PATH='/nix/store/d3aa78jf7pm9lk6ad67x4f21a068dxn6-python-2.7.15/bin:/nix/store/rd4l1qkpx4qfzamh08a701nxr34kgjzr-python2.7-credstash-1.14.0/bin:/nix/store/m4w6rq47p317klmfiixh844jrj6l6wla-python2.7-setuptools-39.2.0/bin:/nix/store/1ahzxgr72vf1za50m34rc3m3vx79n9jq-python2.7-jmespath-0.9.0/bin:/nix/store/ps9mlc8zrd5w10im7pwd6vlfrzxbzbqv-python2.7-docutils-0.14/bin'${PATH:+':'}$PATH
              ^
SyntaxError: invalid syntax
```

bin/.credstash-wrapped
```python
#!/nix/store/d3aa78jf7pm9lk6ad67x4f21a068dxn6-python-2.7.15/bin/python2.7

# -*- coding: utf-8 -*-
import sys;import site;import functools;sys.argv[0] = '/nix/store/rd4l1qkpx4qfzamh08a701nxr34kgjzr-python2.7-credstash-1.14.0/bin/credstash';functools.reduce(lambda k, p: site.addsitedir(p, k), ['/nix/store/rd4l1qkpx4qfzamh08a701nxr34kgjzr-python2.7-credstash-1.14.0/lib/python2.7/site-packages','/nix/store/4lf8as80xw6x6wpbp26grmj5dxvjfd6d-python2.7-idna-2.7/lib/python2.7/site-packages','/nix/store/m4w6rq47p317klmfiixh844jrj6l6wla-python2.7-setuptools-39.2.0/lib/python2.7/site-packages','/nix/store/3js1s9557gx8xhd1a50zyv5zwkzr2kcf-python2.7-asn1crypto-0.24.0/lib/python2.7/site-packages','/nix/store/qm1ajzfan3fzy39yl1wyx7shs693rj3f-python2.7-packaging-17.1/lib/python2.7/site-packages','/nix/store/kimmvi74dnyjbl8i8xwf60z29szmxmh2-python2.7-pyparsing-2.2.0/lib/python2.7/site-packages','/nix/store/v9fvkmqs3j0dw43bhbjmakapc74jdjny-python2.7-six-1.11.0/lib/python2.7/site-packages','/nix/store/ikzm75zqzjvzkfqjsgny638ga1zr8x3y-python2.7-enum34-1.1.6/lib/python2.7/site-packages','/nix/store/9xghys844gq1lvz091m4fn2ds620k2wc-python2.7-ipaddress-1.0.18/lib/python2.7/site-packages','/nix/store/i726wa1h1831xn7i5xvdirjqmwr0mihn-python2.7-pycparser-2.18/lib/python2.7/site-packages','/nix/store/6ph0pq9inpbjm09ad8yrzh3ghar89x56-python2.7-cffi-1.11.5/lib/python2.7/site-packages','/nix/store/655sdqhcc0h3367585qfw9n55b7186jz-python2.7-cryptography-2.2.2/lib/python2.7/site-packages','/nix/store/65gj2v27ry7pkykhlhnpb9r2mqgnw5sd-python2.7-boto3-1.7.48/lib/python2.7/site-packages','/nix/store/mhlgk1jws791kx3hw7gdjhcv1f0ki08h-python2.7-botocore-1.10.48/lib/python2.7/site-packages','/nix/store/g9l0dlgz006fpdvjlnizn5qxr3g574d0-python2.7-python-dateutil-2.7.3/lib/python2.7/site-packages','/nix/store/x24bh0z3wmjpj71bnffldxk3q4qjy8nj-python2.7-setuptools_scm-2.1.0/lib/python2.7/site-packages','/nix/store/1ahzxgr72vf1za50m34rc3m3vx79n9jq-python2.7-jmespath-0.9.0/lib/python2.7/site-packages','/nix/store/lr51nbn3qgwm7aq2pq4x3b63drk9j3ry-python2.7-ply-3.11/lib/python2.7/site-packages','/nix/store/ps9mlc8zrd5w10im7pwd6vlfrzxbzbqv-python2.7-docutils-0.14/lib/python2.7/site-packages','/nix/store/bwl99y7vhh77g695ni9i3yljznvbq9w6-python2.7-ordereddict-1.1/lib/python2.7/site-packages','/nix/store/jjibsz0ypvza72m13rl6k2jkp02i2fmk-python2.7-simplejson-3.15.0/lib/python2.7/site-packages','/nix/store/z9s4p6x8r0r2b3djpyf464fybbmpn2b9-python2.7-s3transfer-0.1.13/lib/python2.7/site-packages','/nix/store/9850dqiinkks72vyb3ysab8wgnj4wa7z-python2.7-futures-3.2.0/lib/python2.7/site-packages','/nix/store/a4samxpliaszak5hs25jad47vqlhharx-python2.7-PyYAML-3.12/lib/python2.7/site-packages'], site._init_pathinfo());
import re
import sys

from credstash import main

if __name__ == '__main__':
    sys.argv[0] = re.sub(r'(-script\.pyw?|\.exe)?$', '', sys.argv[0])
    sys.exit(main())
```

bin/credstash.py
```sh
#! /nix/store/8zkg9ac4s4alzyf4a8kfrig1j73z66dw-bash-4.4-p23/bin/bash -e
export PATH='/nix/store/d3aa78jf7pm9lk6ad67x4f21a068dxn6-python-2.7.15/bin:/nix/store/rd4l1qkpx4qfzamh08a701nxr34kgjzr-python2.7-credstash-1.14.0/bin:/nix/store/m4w6rq47p317klmfiixh844jrj6l6wla-python2.7-setuptools-39.2.0/bin:/nix/store/1ahzxgr72vf1za50m34rc3m3vx79n9jq-python2.7-jmespath-0.9.0/bin:/nix/store/ps9mlc8zrd5w10im7pwd6vlfrzxbzbqv-python2.7-docutils-0.14/bin'${PATH:+':'}$PATH
export PYTHONNOUSERSITE='true'
exec -a "$0" "/nix/store/rd4l1qkpx4qfzamh08a701nxr34kgjzr-python2.7-credstash-1.14.0/bin/.credstash.py-wrapped"  "${extraFlagsArray[@]}" "$@"
```

I've repro'd with the Python 3 version as well.

### Steps to reproduce

```
nix-env --install --attr nixos.python27Packages.credstash
credstash setup
```

## Technical details

 - system: `"x86_64-linux"`
 - host os: `Linux 4.14.54, NixOS, 18.09pre145679.dae9cf6106d (Jellyfish)`
 - multi-user?: `yes`
 - sandbox: `no`
 - version: `nix-env (Nix) 2.0.4`
 - channels(root): `"nixos-18.09pre145679.dae9cf6106d"`
 - nixpkgs: `/nix/var/nix/profiles/per-user/root/channels/nixos/nixpkgs`

I've also repro'd on Ubuntu 18.04

### PR DESCRIPTION
**Pull request by ivanbrennan**
###### Motivation for this change

I want to use the credstash command-line application, but credstash is currently only offered as a Python library.

In order for this to work, I needed to remove the copy of the library that's placed in `$out/bin` and marked executable during the install phase. Other than the patched shebang and executable bit, it's identical to the library that's installed to `$out/lib/python3.7/site-packages`.

Before postFixup has run `wrapPythonPrograms`, bin/ contains two Python files -- credstash and credstash.py -- where `bin/credstash` is the executable you'd expect a user to invoke from the command-line and `bin/credstash.py` contains the credstash module, which `bin/credstash` imports.

After `wrapPythonPrograms` has run, `bin/credstash` is a shell wrapper around the `bin/.credstash-wrapped` python entrypoint, and `bin/credstash.py` is shell wrapper around the `bin/.credstash.py-wrapped` python module. Invoking `bin/credstash` execs `bin/.credstash-wrapped`, which attempts to import the credstash module from `bin/credstash.py` (now a shell script), rather than either `bin/.credstash.py-wrapped` or `lib/python3.7/site-packages/credstash.py`.

This leads to an error:
```
$ credstash get mykey
Traceback (most recent call last):
  File "/nix/store/hk6yma716w6141lcdh509d6qyyi7zm0i-python3.7-credstash-1.15.0/bin/.credstash-wrapped", line 8, in <module>
    from credstash import main
  File "/nix/store/hk6yma716w6141lcdh509d6qyyi7zm0i-python3.7-credstash-1.15.0/bin/credstash.py", line 2
    export PATH='/nix/store/6lm4gi5iv8fbf1b1mm6g3gfnnv63f1gn-python3-3.7.1/bin:/nix/store/hk6yma716w6141lcdh509d6qyyi7zm0i-python3.7-credstash-1.15.0/bin:/nix/store/2n13gf1zdr39ir5dynxlkqndxgy36g08-python3.7-setuptools-40.4.3/bin:/nix/store/mhnqwpa4y1l81zi4cwx989i8h8z9g67l-python3.7-jmespath-0.9.0/bin:/nix/store/qc6q3a2nv4211wyh7q319v6zzd3ab6pc-python3.7-docutils-0.14/bin'${PATH:+':'}$PATH
              ^
SyntaxError: invalid syntax
```

I was able to resolve this by removing `bin/credstash.py` before the postFixup phase has a chance to wrap anything. Now the executable imports the library correctly:
```
(shell wrapper)
 bin/credstash
       │      (python executable)
       └─> bin/.credstash-wrapped
                  │                        (python library)
                  └─> lib/python3.7/site-packages/credstash.py
```

I tried using `dontWrapPythonPrograms` as an alternative solution, but that leads to a different error because runtime dependency lookups fail:
```
$ credstash get mykey
Traceback (most recent call last):
  File "/run/current-system/sw/bin/credstash", line 7, in <module>
    from credstash import main
  File "/nix/store/8rmldlvlv1z1xl7w02dy7f5qhkzdrg8z-python3.7-credstash-1.15.0/bin/credstash.py", line 26, in <module>
    import boto3
ModuleNotFoundError: No module named 'boto3'
```

Fixes #47751

###### Things done

- [ ] Tested using sandboxing ([nix.useSandbox](http://nixos.org/nixos/manual/options.html#opt-nix.useSandbox) on NixOS, or option `sandbox` in [`nix.conf`](http://nixos.org/nix/manual/#sec-conf-file) on non-NixOS)
- Built on platform(s)
   - [x] NixOS
   - [x] macOS
   - [ ] other Linux distributions
- [ ] Tested via one or more NixOS test(s) if existing and applicable for the change (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
- [ ] Tested compilation of all pkgs that depend on this change using `nix-shell -p nox --run "nox-review wip"`
- [ ] Tested execution of all binary files (usually in `./result/bin/`)
- [ ] Determined the impact on package closure size (by running `nix path-info -S` before and after)
- [ ] Assured whether relevant documentation is up to date
- [ ] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/.github/CONTRIBUTING.md).

---

### CODE REVIEW

File: pkgs/development/python-modules/credstash/default.nix [hunk 1]
```diff
@@ -9,6 +9,8 @@ buildPythonPackage rec {
     sha256 = "814560f99ae2409e2c6d906d878f9dadada5d1d0a950aafb6b2c0d535291bdfb";
   };
 
+  postInstall = "rm $out/bin/credstash.py";
```
Comment (dotlambda) [inline pkgs/development/python-modules/credstash/default.nix]: Please add a comment explaining this.

**Comment by dotlambda** [comment]
What is not clear to me is why this is even in `pythonPackages`? Isn't this only ever used on the command line nd never via Python's `import`?

**Comment by sean-bennett112** [comment]
@dotlambda `credstash` can also be used via Python's `import`. We're currently using it at work in this way.

**Comment by ivanbrennan** [comment]
Same here. We have some internal tools that use the library to interact with credstash.

**Pull request merged**